app/controllers/users_controller.rb

```diff
@@ -2,24 +2,27 @@ class UsersController < ApplicationController
   include Accessible
 
   def new
+    @user = User.new
+    @person = Person.new
   end
 
   def create
-    ActiveRecord::Base.transaction do
-      @person = Person.new(email: user_params[:email])
-      @user = User.new(password: user_params[:password], password_confirmation: user_params[:password])
-      @person.personable = @user
-
-      respond_to do |format|
-        if @person.save
-          reset_session
-          session[:current_user_id] = @user.id
-          format.html { redirect_to dashboard_path, notice: "Account was successfully created." }
-          format.json { render :show, status: :created, location: @user }
-        else
-          format.html { render :new }
-          format.json { render json: @user.errors, status: :unprocessable_entity }
-        end
+    context = RegistrationContext.new params.permit(:email), params.permit(:password)
```

Comment:
I don't know about this pattern. It feels like it obfuscates and buries some important business logic into this little pocket on the side, whereas I think it deserves to be more mainline. Rails has a really great built in solution for where to put this kind of logic: in the model concerns & model callbacks. I just found those scale a lot better over time with fewer tradeoffs. Admittedly, I don't think I did the best job of that in the early days of the lava codebase, but the more I leaned into that pattern, it really worked well.

I get that the controller was expanding in scope, and you don't want to bloat it even more with additional extra logic. I definitely would share that goal. But I'm not sure about creating an extra abstraction in order to handle this. I once did a project that involved a lot of service objects of different kinds and it ends up spreading out a lot of important business logic into a bunch of different places and I found it makes it harder to get a unified view of what's happening.

I'd push this into the model. To play this out:

This line could simply be changed to:
```
@user = User.new(password: user_params[:password], password_confirmation: user_params[:password])
```

Then within the User model create a concern such as `include Registerable` where all this RegistrationContext lives. It can do all the things that Context is doing.

A potential worry of adding all this logic into the model is that the model will collect quite a lot of logic over time, but the combination of Concerns and callbacks make that quite easy to manage LOTS of business logic all wrapped behind the ActiveRecord API for models. I think it's actually not hard to manage, provided we keep the file organized, and we keep bundling things into concerns.

I've been going through 37signals basecamp source code snippets and many of their models are huge, but they're surprisingly easy to wrangle still. Here's one example: https://github.com/krschacht/37signals-rails-code/blob/main/basecamp/app/models/recording.rb

I think the two main advantages of this are:

* You open a model like the one I linked to above and the whole thing reads almost like a table of contents! You start with a list of concerns that name large chunks of logic, followed by core associations and hooks.

* There will inevitably be other places that we create User objects and the vast majority of the time we're also going to want to fully initialize those with a Person record and other stuff. Having it in the model encourages that golden path, and the one-off times we don't want to we can use the various blocks that activerecord exposes to suppress callbacks. Whereas when we sprinkle this business logic in Contexts, it's pushing us to create a new Context for a slightly different situation and duplicate a lot of the same logic.

* There are often bits of logic that we are adding to one model which initially seem like they apply onto to that model: so we start off with a concern that is used solely by that model. But over time we see ourselves doing something similar in another model. After the 2nd or 3rd time we generalize a single concern and mix it in for all the models that do it. So starting off with concerns really sets you up for lifting that logic up and sharing it across models. And the same applies to hooks as well. You start off with one-off hooks, realize you're doing this in a few places and generalize and elevate that hook up.

What do you think about pushing the models further before creating new classes?

Reply:
Oooh interesting. I didn't consider a model concern, I've always exclusively used those for features which are shared between multiple models (Taggable, Commentable, etc) I'll rearrange this code and see what I come up with. Good suggestion!

I proposed this because it's a pattern I've seen which did better than the kind of lazy idea of just shoving everything into one file -- but there's a lot in there which isn't great.

Reply:
Nice. This is probably more info than you're interested in so don't feel obligated to consume this :) but a year ago I stumbled on a bunch of talks that DHH gave really unpacking how he thinks about organizing rails code. This is what started me studying the Basecamp code more closely. It changed a lot of my thinking. I think these two are the ones related to what we're going back and forth on, and were probably the most interesting ones, all around:

https://www.youtube.com/watch?v=M3JPTOTqsnE
https://www.youtube.com/watch?v=Yd3j3ZEx6DU

Reply:
Ah yes, I'm remembering a bit more. One of the things about the registration action is that it's not constrained to a single object. Because we have Person(:email) and User(:password), the registration form gets weird.
Rendering errors really gets weird, which is why I spent so much of this code wrangling errors into a single Errors list.

It also doesn't really play well with StrongParameters, or at least I don't know how to make them play well together.

I'm going to find some more docs on delegated types and see if there's an easy way through this it.

Reply:
@krschacht this worked out brilliantly. I need to move the tests around still, but I'm much happier with the code organization using nested attributes. Thanks for your nudge.